File: models/question.py
import sqlite3
from dataclasses import dataclass
from typing import Optional
from datetime import datetime as dt

from db import connection


@dataclass
class Question:
    question: str
    answer: str
    accepted_answer: str
    comment: str
    source: str
    handout_str: str
    handout_img: str
    got_questions_id: int = None
    author_id: Optional[int] = None
    package_id: Optional[int] = None
    difficulty: Optional[float] = None
    is_incorrect: bool = None
    id: int = None

    @classmethod
    def build_question(cls, question_dict):
        author_id = (question_dict["authors"] or [{"id": None}])[0]["id"]

        return cls(
            got_questions_id=question_dict["id"],
            question=question_dict["text"],
            answer=question_dict["answer"],
            accepted_answer=question_dict["zachet"],
            comment=question_dict["comment"],
            handout_str=question_dict["razdatkaText"],
            handout_img=question_dict["razdatkaPic"],
            source=question_dict["source"],
            author_id=author_id,
            package_id=question_dict.get("packId", -1),
            difficulty=question_dict["complexity"],
            is_incorrect=question_dict["takenDown"],
        )

    @classmethod
    def has_questions_from_package(cls, package_id):
        with connection() as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT COUNT(*) FROM questions WHERE package_id = ?", (package_id,))
            return cursor.fetchone()[0] > 0

    @classmethod
    def delete_all_questions_for_package(cls, package_id):
        with connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                DELETE FROM images 
                WHERE question_id IN (SELECT id FROM questions WHERE package_id = ?)
            """,
                (package_id,),
            )
            cursor.execute("DELETE FROM questions WHERE package_id = ?", (package_id,))
            conn.commit()

    @classmethod
    def question_ids_for_year(cls, year):
        start_timestamp = dt.timestamp(dt(year, 1, 1))
        end_timestamp = dt.timestamp(dt(year + 1, 1, 1))
        with connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                SELECT id FROM questions 
                WHERE package_id IN (
                    SELECT gotquestions_id FROM packages WHERE end_date between ? and ?
                )
                AND is_incorrect = 0
            """,
                (start_timestamp, end_timestamp),
            )
            return [row[0] for row in cursor.fetchall()]

    @classmethod
    def find(cls, ids: list[int]) -> list["Question"]:
        with connection() as conn:
            cursor = conn.cursor()
            cursor.row_factory = sqlite3.Row
            rows = cursor.execute(
                f"""
                select q.id, q.question, q.answer, q.comment, q.accepted_answer, q.handout_str, q.source, 
                    i.mime_type, i.data as image_data 
                from questions q
                left join images i on q.id = i.question_id
                where q.id in ({','.join('?' * len(ids))})
            """,
                ids,
            ).fetchall()
            return [cls(
            id=row["id"],
            question=row["question"],
            answer=row["answer"],
            accepted_answer=row["accepted_answer"],
            comment=row["comment"],
            handout_str=row["handout_str"],
            handout_img=row["image_data"],
            source=row["source"],
        ) for row in rows]

File: internal/models/question.go
package models

import (
	"database/sql"
	"fmt"
	"questions-vote/internal/db"
	"strings"
)

// QuestionRepository handles question database operations
type QuestionRepository struct {
	db *sql.DB
}

// NewQuestionRepository creates a new question repository
func NewQuestionRepository() *QuestionRepository {
	return &QuestionRepository{
		db: db.GetDB(),
	}
}

// FindByIDs retrieves questions by their IDs
func (r *QuestionRepository) FindByIDs(ids []int) ([]*Question, error) {
	if len(ids) == 0 {
		return []*Question{}, nil
	}
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "?"
		args[i] = id
	}

	query := fmt.Sprintf(`
		SELECT q.id, q.question, q.answer, q.comment, q.accepted_answer, 
		       q.handout_str, q.source, i.data as image_data 
		FROM questions q
		LEFT JOIN images i ON q.id = i.question_id
		WHERE q.id IN (%s)
	`, strings.Join(placeholders, ","))

	rows, err := r.db.Query(query, args...)
	if err != nil {
		return nil, fmt.Errorf("failed to query questions: %w", err)
	}
	defer rows.Close()

	var questions []*Question
	for rows.Next() {
		q := &Question{}
		var imageData sql.NullString
		err := rows.Scan(
			&q.ID,
			&q.Question,
			&q.Answer,
			&q.Comment,
			&q.AcceptedAnswer,
			&q.HandoutStr,
			&q.Source,
			&imageData,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to scan question: %w", err)
		}

		if imageData.Valid {
			q.HandoutImg = imageData.String
		}

		questions = append(questions, q)
	}

	if err = rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating over rows: %w", err)
	}

	return questions, nil
}

// GetQuestionIDsForYear returns question IDs for a specific year
func (r *QuestionRepository) GetQuestionIDsForYear(year int) ([]int, error) {
	startTimestamp := int64(year * 365 * 24 * 3600)
	endTimestamp := int64((year + 1) * 365 * 24 * 3600)

	query := `
		SELECT id FROM questions 
		WHERE package_id IN (
			SELECT gotquestions_id FROM packages 
			WHERE end_date BETWEEN ? AND ?
		)
		AND is_incorrect = 0
	`

	rows, err := r.db.Query(query, startTimestamp, endTimestamp)
	if err != nil {
		return nil, fmt.Errorf("failed to query question IDs: %w", err)
	}
	defer rows.Close()

	var ids []int
	for rows.Next() {
		var id int
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("failed to scan question ID: %w", err)
		}
		ids = append(ids, id)
	}

	return ids, rows.Err()
}

// HasQuestionsFromPackage checks if a package has any questions
func (r *QuestionRepository) HasQuestionsFromPackage(packageID int) (bool, error) {
	query := `SELECT 1 FROM questions WHERE package_id = ? LIMIT 1`

	var exists int
	err := r.db.QueryRow(query, packageID).Scan(&exists)
	if err != nil {
		if err == sql.ErrNoRows {
			return false, nil
		}
		return false, fmt.Errorf("failed to check if package has questions: %w", err)
	}

	return true, nil
}

// BuildQuestionFromDict creates a question from API response data
func BuildQuestionFromDict(questionDict map[string]interface{}, packageID int) (*Question, error) {
	id, ok := questionDict["id"].(float64)
	if !ok {
		return nil, fmt.Errorf("invalid id in question data")
	}

	text, ok := questionDict["text"].(string)
	if !ok {
		return nil, fmt.Errorf("invalid text in question data")
	}

	answer, ok := questionDict["answer"].(string)
	if !ok {
		return nil, fmt.Errorf("invalid answer in question data")
	}

	var zachet, comment, razdatkaText, razdatkaPic, source string
	var complexity float64
	var takenDown bool
	var authorID *int

	if val, exists := questionDict["zachet"]; exists && val != nil {
		zachet, _ = val.(string)
	}

	if val, exists := questionDict["comment"]; exists && val != nil {
		comment, _ = val.(string)
	}

	if val, exists := questionDict["razdatkaText"]; exists && val != nil {
		razdatkaText, _ = val.(string)
	}

	if val, exists := questionDict["razdatkaPic"]; exists && val != nil {
		razdatkaPic, _ = val.(string)
	}

	if val, exists := questionDict["source"]; exists && val != nil {
		source, _ = val.(string)
	}

	if val, exists := questionDict["complexity"]; exists && val != nil {
		complexity, _ = val.(float64)
	}

	if val, exists := questionDict["takenDown"]; exists && val != nil {
		takenDown, _ = val.(bool)
	}

	if authorsInterface, exists := questionDict["authors"]; exists && authorsInterface != nil {
		if authors, ok := authorsInterface.([]any); ok && len(authors) > 0 {
			if author, ok := authors[0].(map[string]any); ok {
				if idVal, exists := author["id"]; exists && idVal != nil {
					if idFloat, ok := idVal.(float64); ok {
						id := int(idFloat)
						authorID = &id
					}
				}
			}
		}
	}

	return &Question{
		GotQuestionsID: int(id),
		Question:       text,
		Answer:         answer,
		AcceptedAnswer: zachet,
		Comment:        comment,
		HandoutStr:     razdatkaText,
		HandoutImg:     razdatkaPic,
		Source:         source,
		AuthorID:       authorID,
		PackageID:      &packageID,
		Difficulty:     &complexity,
		IsIncorrect:    &takenDown,
	}, nil
}

// GetQuestionIDsFromPackage returns question IDs for a specific package
func (r *QuestionRepository) GetQuestionIDsFromPackage(packageID int) ([]int, error) {
	query := `SELECT id FROM questions WHERE package_id = ?`

	rows, err := r.db.Query(query, packageID)
	if err != nil {
		return nil, fmt.Errorf("failed to query questions: %w", err)
	}
	defer rows.Close()

	var questionIDs []int
	for rows.Next() {
		var id int
		err := rows.Scan(&id)
		if err != nil {
			return nil, fmt.Errorf("failed to scan question ID: %w", err)
		}
		questionIDs = append(questionIDs, id)
	}

	return questionIDs, rows.Err()
}
